core/src/main/java/org/apache/iceberg/io/SortedPosDeleteWriter.java

```diff
@@ -0,0 +1,190 @@
+
+  private final FileAppenderFactory<T> appenderFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileFormat format;
+  private final PartitionKey partition;
+  private final long recordsNumThreshold;
+
+  private int records = 0;
+
+  SortedPosDeleteWriter(FileAppenderFactory<T> appenderFactory,
+                        OutputFileFactory fileFactory,
+                        FileFormat format,
+                        PartitionKey partition,
+                        long recordsNumThreshold) {
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.format = format;
+    this.partition = partition;
+    this.recordsNumThreshold = recordsNumThreshold;
+  }
+
+  SortedPosDeleteWriter(FileAppenderFactory<T> appenderFactory,
+                        OutputFileFactory fileFactory,
+                        FileFormat format,
+                        PartitionKey partition) {
+    this(appenderFactory, fileFactory, format, partition, DEFAULT_RECORDS_NUM_THRESHOLD);
+  }
+
+  public void delete(CharSequence path, long pos) {
+    delete(path, pos, null);
+  }
+
+  public void delete(CharSequence path, long pos, T row) {
+    posDeletes.compute(CharSequenceWrapper.wrap(path), (k, v) -> {
+      if (v == null) {
+        return Lists.newArrayList(PosValue.of(pos, row));
+      } else {
+        v.add(PosValue.of(pos, row));
+        return v;
+      }
+    });
+
+    records += 1;
+
+    // TODO Flush buffer based on the policy that checking whether whole heap memory size exceed the threshold.
+    if (records >= recordsNumThreshold) {
+      flushBuffer();
+    }
+  }
+
+  public List<DeleteFile> complete() throws IOException {
+    close();
+
+    return completedFiles;
+  }
+
+  public Set<CharSequence> referencedDataFiles() {
+    return referencedDataFiles;
+  }
+
+  @Override
+  public void close() throws IOException {
+    flushBuffer();
+  }
+
+  private void flushBuffer() {
+    if (posDeletes.isEmpty()) {
+      return;
+    }
+
+    // Create a new output file.
+    EncryptedOutputFile outputFile;
+    if (partition == null) {
+      outputFile = fileFactory.newOutputFile();
+    } else {
+      outputFile = fileFactory.newOutputFile(partition);
+    }
+
+    PositionDeleteWriter<T> writer = appenderFactory.newPosDeleteWriter(outputFile, format, partition);
+    try (PositionDeleteWriter<T> closeableWriter = writer) {
+      // Sort all the paths.
+      CharSequence[] paths = new CharSequence[posDeletes.size()];
+      int index = 0;
+      for (CharSequenceWrapper charSequenceWrapper : posDeletes.keySet()) {
+        paths[index] = charSequenceWrapper.get();
+        index += 1;
+      }
+      Arrays.sort(paths, Comparators.charSequences());
+
+      // Write all the sorted <path, pos, row> triples.
+      for (CharSequence path : paths) {
+        List<PosValue<T>> positions = posDeletes.get(wrapper.set(path));
+        positions.sort(posValueComparator);
```

Comment:
Nit: could probably be `positions.sort(Comparator.comparingLong(PosValue::pos))`

Reply:
That sounds good.